src/identity.js

```diff
@@ -436,93 +436,77 @@ export class Identity extends EventEmitter {
+                if (!sessionData && !this._itpMode) {
+                    sessionData = await this._hasSession.get('rpc/hasSession.js', { autologin: autoLoginConverted });
                 }
-                if (this._itpMode || (isObject(data.error) && data.error.type === 'LoginException')) {
-                    data = await this._spid.get('ajax/hasSession.js', { autologin: autoLoginConverted });
+                if (this._itpMode || (sessionData && isObject(sessionData.error) && sessionData.error.type === 'LoginException')) {
+                    sessionData = await this._spid.get('ajax/hasSession.js', { autologin: autoLoginConverted });
                 }
 
-                const shouldShowItpModal = this._itpModalRequired() && !this._itpMode &&
-                    isObject(data.error) && data.error.type === 'UserException' &&
-                    this.cache.get(LOGIN_IN_PROGRESS_KEY) !== null;
+                const shouldShowItpModal = this._itpModalRequired() && !this._itpMode && sessionData && isObject(sessionData.error)
+                    && sessionData.error.type === 'UserException' && this.cache.get(LOGIN_IN_PROGRESS_KEY) !== null;
                 if (shouldShowItpModal) {
                     this.cache.delete(LOGIN_IN_PROGRESS_KEY);
                     const modal = new ItpModal(this._spid, this.clientId, this.redirectUri, this.env);
-                    data = await modal.show();
+                    sessionData = await modal.show()
                 }
-                if (this._enableSessionCaching) {
-                    const expiresIn = 1000 * (data.expiresIn || 300);
-                    this.cache.set(HAS_SESSION_CACHE_KEY, data, expiresIn);
+                if (sessionData && this._enableSessionCaching) {
+                    const expiresIn = 1000 * (sessionData.expiresIn || 300);
+                    this.cache.set(HAS_SESSION_CACHE_KEY, sessionData, expiresIn);
                 }
-                postProcess(data);
-                this._session = data;
-                resolve(data);
-            } catch (err) {
-                this.emit('error', err);
-                reject(new SDKError('HasSession failed', err));
-            }
-        };
-        const promise = new Promise(promiseFn);
-        this._hasSessionInProgress = promise.then(v => {
-            this._hasSessionInProgress = null;
-            return v;
-        }, e => {
-            this._hasSessionInProgress = null;
-            throw e;
-        });
+                _postProcess(sessionData);
+                this._session = sessionData;
+                return sessionData;
+            };
+            this._hasSessionInProgress = _getSession()
+                .catch(err => {
+                    this.emit('error', err);
+                    throw new SDKError('HasSession failed', err);
+                })
+                .finally(() => {
+                    this._hasSessionInProgress = false;
+                });
```

Comment:
Given that we already ask people to polyfill `Promise`, I'm hoping that this change will be kosher? It's currently in stage 4 and has quite good support:

![Promise_prototype_finally___-_JavaScript___MDN](https://user-images.githubusercontent.com/463105/67126184-c91ed280-f1f6-11e9-90c0-c919aaae3761.jpg)

If not, we can always revert it back to the existing solution